File: 프로그래머스/2/12909. 올바른 괄호/올바른 괄호.cpp
#include<string>
#include<iostream>
#include<stack>
using namespace std;

bool solution(string s)
{
    bool answer = false;
    stack<int> st;
    
    for(int i = 0; i < s.length(); i++){
        // '('일 때는 무조건 push
        if(s[i] == '(') st.push(s[i]);
        else {
            // ')'일 때
            // 스택이 비어있으면 바로 false
            if(st.empty()) return false;
            else {
                // 스택이 비어있지않을 때
                // '(' 인 경우 pop
                if(st.top() == '('){
                    st.pop();
                } else {
                    st.push(s[i]);
                }
            }
        }
    }
    
    if(st.empty()){
        answer = true;
    }
    
    return answer;
}

File: 프로그래머스/2/12909. 올바른 괄호/올바른 괄호.java
import java.util.*;

class Solution {
    boolean solution(String s) {
        boolean answer = true;
        Stack <Character> st = new Stack<>();
        
        for(int i = 0; i < s.length(); i++){
            char c = s.charAt(i); // 하나 하나씩 선택
            
            if(c == '('){
                st.push(c);
            } else {
                // ')' 일 때 비어있지 않으면
                if(!st.isEmpty()){
                    if(st.peek() == '('){
                        st.pop();
                    } else {
                        st.push(c);
                    }
                } else {
                    // 비어있으면 바로 false
                    answer = false;
                    break;
                }
            }
        }
        
        if(!st.isEmpty()){
            answer = false;
        }

        return answer;
    }
}